File: Swift/Sort/[백준 1181] 단어 정렬.swift
//
//  [백준 1181] 단어 정렬.swift
//  Algorithm
//
//  Created by 김민 on 2/12/24.
//
// 백준 1181 단어 정렬 https://www.acmicpc.net/problem/1181

/*
[📌 전략]
중복된 값을 입력받지 않기 위해 set collection을 사용했다.
swift의 sort 알고리즘은 stable sort이므로 정렬을 진행했을 때, 중복된 값의 순서가 바뀌지 않음을 의미한다.
따라서 먼저 사전순으로 정렬해 준 뒤, 길이가 짧은 순으로 다시 한번 정렬했다.

[📝 기록]
set collection은 순서가 없으니 막연하게 정렬이 안 될 것이라 생각하고, 처음에 배열로 바꿔서 정렬했는데
Set<String>은 sorted 사용이 가능하고 이는 [String]을 반환한다.
*/

let n = Int(readLine()!)!
var inputs = Set<String>()

for _ in 0..<n {
    inputs.insert(readLine()!)
}

var words = inputs.sorted()
words.sort { $0.count < $1.count }

/*
 var words = Array(inputs)
 words.sort()
 words.sort { $0.count < $1.count }
 */

print(words.joined(separator: "\n"))

File: Python/정렬/[백준 1181] 단어 정렬.py
'''
[백준 1181] 단어 정렬
https://www.acmicpc.net/problem/1181

[📌 전략]
- 기준이 여러 개인 정렬 문제

[✅ 풀이]
1. set을 선언하고, 입력받은 단어를 set에 추가
2. key=lambda 사용하여 길이가 짧은 것부터, 길이가 같으면 사전순으로 정렬
3. 출력

[📝 기록]
LV: 실버5
풀이 시간: 7분
'''

import sys

input = sys.stdin.readline 

n = int(input())
words = set() # 중복될 수 없으므로 set으로 처리

for _ in range(n): words.add(input().strip()) # 공백 제거 후 set에 추가

sorted_words = sorted(words, key=lambda x: (len(x), x)) # 길이가 짧은 것부터, 길이가 같으면 사전 순으로

print(*sorted_words, sep='\n')